content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,242 @@
+
+The example story is, we have some books and want to pass the author name as a parameter, and get the book title, author name, and publish date as a summary.
+```ruby
+class Book
+  attr_reader :title, :publish_year
+ 
+  def initialize(title, publish_year)
+    @title = title
+    @publish_year = publish_year
+  end
+ 
+  def summary(first_name, last_name)
+    "#{title} - #{last_name}, #{first_name} (#{publish_year})"
+  end
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary("Bater", "Chen")
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+We have a book class, it has two attributes, title, and publish_year. The `initialize` method would be called when we new a book object. Book class only has one instance method called `summary`, it needs two parameters, first_name and last_name. In Ruby, when we define the `attr_reader` as the attribute of the object, it will create the read method automatically. So we can call the instance variable directly by `title` to get `@title` return, same as publish_year.
+
+This simple method works fine but is not good enough. Now we want to do some refactoring on it.
+
+## The good code should be small and simple
+What is the good code? This question might be hard to answer, but I would say **the good code should be small and simple.** In this article, let's focus on making the parameter smaller.
+
+Generally speaking, **one parameter is better than two, and no parameter is better than one.** The fact of the parameter is, it cost time to read and understand the business logic, and time is always expensive in all projects. We spent around 70% of the time reading the existing code, but only 30% is actually writing. If we can save some time from reading the code, not only from literally reading but also from understanding the spec and logic, which means we can develop efficiently.
+
+Secondly, when **the code is short, the bugs are nowhere to hide.** Parameter causes bugs easily because it needs extra concern. Reducing parameters can make the method more clear, and show the intention explicitly. It can help us to maintain the code easily. When we try to reduce the parameter, we are actually trying to reduce the potential bugs at the same time.
+
+Now we know it's better to reduce the parameter, but how to do it? Let's check the parameter again.
+
+## Data clump
+When we find out the parameters are **a group of data that always come together**, and it would make no sense when missing one of them, we can call it a set of **Data Clump**. For example, a pair of start date and end date, or in this example, first name and last name.
+
+In the OOP world, the **object is the container of data, and class is the container of method** (behavior of object). It's cheap to create many small objects to manage and operate the serial data. Additionally, we can define the method on the object and reuse it when necessary.
```

Comment:
[grammar] shouldn't need a `,` when you join just two clauses with `and`

```suggestion
In the OOP world, the **object is the container of data and class is the container of methods** (i.e. the behavior of objects). It's cheap to create many small objects to manage and operate on serial data. Additionally, we can define methods on the object and reuse them when necessary.
```

Reply:
> I think the right approach is a footnote explaining that:
> 
> 1. Many small objects could end up causing GC pressure.
> 2. But high performance languages often have tricks up their sleeve like doingg [escape analysis](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/) to convert heap allocation to stack allocations or [expanding argument objects as primitive arguments](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/). These languages also generally have separate GC pathways for small objects and short-lived objects. In most cases, if you're just using an argument object locally in a high performance language, your runtime will manage to optimize most of the cost away.

Hi @boazy , thanks for your feedback.

The link for `escape analysis ` and `expanding argument objects as primitive arguments ` are the same, is that your intention?